<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>The scvis model · scVI</title><script data-outdated-warner src="assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.045/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.4/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.24/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="index.html"><img src="assets/logo.svg" alt="scVI logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="index.html">scVI</a></span></div><form class="docs-search" action="search.html"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="index.html">Getting started</a></li><li><a class="tocitem" href="DataProcessing.html">Data processing</a></li><li><a class="tocitem" href="scVAE.html">The scVAE model</a></li><li><a class="tocitem" href="scLDVAE.html">The scLDVAE model</a></li><li><a class="tocitem" href="Training.html">Model training</a></li><li><a class="tocitem" href="Evaluation.html">Model evaluation</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href="scvis.html">The scvis model</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href="scvis.html">The scvis model</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/maren-ha/scVI.jl/blob/origin/main/docs/src/scvis.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="The-scvis-model"><a class="docs-heading-anchor" href="#The-scvis-model">The scvis model</a><a id="The-scvis-model-1"></a><a class="docs-heading-anchor-permalink" href="#The-scvis-model" title="Permalink"></a></h1><p>scVAE model with additional tSNE loss term, based on the scvis model proposed in <a href="https://doi.org/10.1038/s41467-018-04368-5">Ding J, Condon A and Shah SP Interpretable dimensionality reduction of single cell transcriptome data with deep generative models. <em>Nat Commun</em> <strong>9</strong>, 2002 (2018)</a>.</p><p>From the abstract: </p><blockquote><p>scvis is a statistical model that captures and visualizes the low-dimensional structures in single-cell gene expression data. It is robust to the number of data points and learns a probabilistic parametric mapping function to add new data points to an existing embedding.</p></blockquote><h2 id="The-additional-t-SNE-loss-component"><a class="docs-heading-anchor" href="#The-additional-t-SNE-loss-component">The additional t-SNE loss component</a><a id="The-additional-t-SNE-loss-component-1"></a><a class="docs-heading-anchor-permalink" href="#The-additional-t-SNE-loss-component" title="Permalink"></a></h2><p>The model is based on adding a t-SNE objective to the VAE loss function (i.e., the ELBO). This component is supposed to help structure the latent representation to look more like t-SNE. </p><p>Similar to the usual t-SNE objective, a matrix transition of transition probabilities has to be calculated. For this, individual perplexities have to be calculated. </p><article class="docstring"><header><a class="docstring-binding" id="scVI.scvis.compute_transition_probs" href="#scVI.scvis.compute_transition_probs"><code>scVI.scvis.compute_transition_probs</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Computes the transition probabilities between datapoints in X using the perplexity based similarity approach.</p><p>Parameters:</p><ul><li>X: The data matrix with shape obs x vars.</li><li>perplexity: The perplexity value to use for computing the transition probabilities. Default is 30.</li></ul><p>Returns:</p><ul><li>A matrix of transition probabilities with shape obs x obs, where P[i,j] represents the transition probability from  datapoint i to j.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maren-ha/scVI.jl/blob/00140c4ceb38b8380db2e6d2ab5e2cfc756bd82b/src/scvis/scvis_core.jl#L101-L111">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="scVI.scvis.compute_differentiable_transition_probs" href="#scVI.scvis.compute_differentiable_transition_probs"><code>scVI.scvis.compute_differentiable_transition_probs</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Compute differentiable transition probabilities from a given matrix X.
Parameters:
-----------
X : AbstractMatrix{S}
    The matrix from which to compute the differentiable transition probabilities.
perplexity : S
    The perplexity to use for the computation of the transition probabilities.
Returns:
--------
P : matrix of type S
    The differentiable transition probabilities.</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maren-ha/scVI.jl/blob/00140c4ceb38b8380db2e6d2ab5e2cfc756bd82b/src/scvis/scvis_diff.jl#L82-L94">source</a></section></article><p>Based on these transition probabilities, the t-SNE loss component is calculated and integrated with the standard ELBO. </p><article class="docstring"><header><a class="docstring-binding" id="scVI.scvis.tsne_repel" href="#scVI.scvis.tsne_repel"><code>scVI.scvis.tsne_repel</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Compute the t-SNE cost function with repulsion term.</p><p>Args:</p><ul><li>z: matrix of shape (latent_dim, batchsize) representing the low-dimensional   embedding of the data.</li><li>P: matrix of shape (batchsize, batchsize) representing the symmetric    pairwise affinities of the data points.</li></ul><p>Returns:</p><ul><li>The t-SNE cost function value normalized by the batchsize.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maren-ha/scVI.jl/blob/00140c4ceb38b8380db2e6d2ab5e2cfc756bd82b/src/scvis/scvis_core.jl#L119-L130">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="scVI.scvis.tsne_loss" href="#scVI.scvis.tsne_loss"><code>scVI.scvis.tsne_loss</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Computes the t-SNE loss for the given scVAE model, tsne network, input data, and transition probabilities.</p><p>Parameters:     - m: scVAE model     - tsne_net: Dense network for t-SNE     - x: Input data, shape obs x vars      - P: Transition probabilities, shape obs x obs, defaults to <code>Nothing</code>     - epoch: Current training epoch</p><p>Returns:     - kl_qp: t-SNE loss</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maren-ha/scVI.jl/blob/00140c4ceb38b8380db2e6d2ab5e2cfc756bd82b/src/scvis/scvis_diff.jl#L102-L114">source</a></section></article><article class="docstring"><header><a class="docstring-binding" id="scVI.scvis.scvis_loss" href="#scVI.scvis.scvis_loss"><code>scVI.scvis.scvis_loss</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">Computes the loss value of the scVI model.

It takes in an scVAE model, a matrix of input data (x), and an optional matrix of transition probabilities (P).
It applies the encoder to the input data to obtain a set of parameters for the approximate posterior of the latent variable (z).
It then uses these parameters to perform the reparameterization trick to sample from the approximate posterior.
It then applies the generative model to the sampled latent variable to obtain the parameters of the likelihood.
It then computes the KL divergence between the approximate posterior and the prior, and the reconstruction loss between the input data and the likelihood.
It also computes the tsne_repel loss which is the repulsion loss between the latent representations of the data.
Finally, it returns the sum of reconstruction loss, KL divergence, and tsne_repel loss normalized by the batch size.

Parameters:
- m: an instance of the scVAE model
- x: the input data, a matrix of shape obs x vars
- P: (Optional) the transition probabilities matrix, defaults to nothing
- kl_weight: (Optional) weight for the KL divergence term, defaults to 1.0
- epoch: (Optional) current training epoch, defaults to 1
Returns:
- lossval: the loss value

Parameters:
- m: an instance of the scVAE model
- x: input data of shape obs x vars
- P: (Optional) matrix of transition probabilities. If not provided, it will be computed internally.
- kl_weight: weighting factor for KL divergence. 
- epoch: The current epoch number.

Returns:
- lossval: The final loss value</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maren-ha/scVI.jl/blob/00140c4ceb38b8380db2e6d2ab5e2cfc756bd82b/src/scvis/scvis_core.jl#L151-L180">source</a></section></article><p>There are different versions how this can be used to train scvis models. </p><article class="docstring"><header><a class="docstring-binding" id="scVI.scvis.train_scvis_model!" href="#scVI.scvis.train_scvis_model!"><code>scVI.scvis.train_scvis_model!</code></a> — <span class="docstring-category">Function</span></header><section><div><p>Trains an scVAE model with the given AnnData, TrainingArgs and mode.</p><p>Parameters:</p><ul><li>m::scVAE: the scVAE model to be trained</li><li>adata::AnnData: the AnnData object containing the dataset</li><li>training_args::TrainingArgs: the TrainingArgs object containing the training parameters such as batch size, learning rate, etc.</li><li>mode::Symbol: the training mode, can be one of the following:   <code>:alternating</code>   <code>:alternating_diff</code>   <code>:joint</code>   <code>:joint_diff</code></li></ul><p>Returns:</p><ul><li>Nothing. The model is trained in place.</li></ul><p>Raises:</p><ul><li>A warning if an unsupported training mode is selected.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maren-ha/scVI.jl/blob/00140c4ceb38b8380db2e6d2ab5e2cfc756bd82b/src/scvis/scvis_train.jl#L2-L20">source</a></section><section><div><p>Train an scVI model in joint mode.</p><p>Parameters:</p><ul><li>m: scVAE model to be trained</li><li>adata: AnnData object containing the data to be used for training</li><li>training_args: TrainingArgs object containing the training hyperparameters, such as weight decay, learning rate, batch size, and maximum number of epochs</li><li>::Val{:joint}: Symbol specifying the mode of training, must be :joint</li></ul><p>Returns:</p><ul><li>Tuple of trained scVAE model, AnnData object</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/maren-ha/scVI.jl/blob/00140c4ceb38b8380db2e6d2ab5e2cfc756bd82b/src/scvis/scvis_train.jl#L72-L84">source</a></section></article></article><nav class="docs-footer"><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.24 on <span class="colophon-date" title="Thursday 20 April 2023 05:39">Thursday 20 April 2023</span>. Using Julia version 1.7.3.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
